### PR TITLE
add fixes for 'productChecksum', 'productsEquals', lazy cart creation, qty for all product types

### LINKED ISSUE #4123: Discount code disappears at the time of placing the order
## Current behavior

After registering on the /checkout page at the time of placing the order, discount code disappears

## Steps to reproduce the issue

1 Create an account on Personal Details step
2 Place the order

## Can you handle fixing this bug by yourself?

- [ ] YES
- [ ] NO

## Which [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html) state this refers to? Info for developer.
Pick one option.

- [ ] This is a bug report for test version on https://test.storefrontcloud.io - In this case Developer should create branch from `develop` branch and create Pull Request `2. Feature / Improvement` back to `develop`.
- [x ] This is a bug report for current Release Candidate version on https://next.storefrontcloud.io - In this case Developer should create branch from `release` branch and create Pull Request `3. Stabilisation fix` back to `release`.
- [ ] This is a bug report for current Stable version on https://demo.storefrontcloud.io and should be placed in next stable version hotfix - In this case Developer should create branch from `hotfix` or `master` branch and create Pull Request `4. Hotfix` back to `hotfix`.

## Environment details
Browser: Google Chrome Version 81.0.4044.17
OS: Win 10
Node: 10.15.2
Code Version: 1.11.1

### PR DESCRIPTION
**Pull request by gibkigonzo**
### Related Issues

closes #4123, #4118, #4104
- Group 'productChecksum' and 'productsEquals' logic for all supported products types.
- Remove 'checksum' when editing product.
- Remove and add coupon when user login
- Remove 'NA' as default company
- Show qty in microcart for all types of product
- Remove preload font - it gives good performance, but vue-meta refresh page, because there is script `onload`.

### Short Description and Why It's Useful

Zrzut ekranu z 2020-02-27 10-32-28

### Which Environment This Relates To
Check your case. In case of any doubts please read about [Release Cycle](https://docs.vuestorefront.io/guide/basics/release-cycle.html)

- [ ] Test version (https://test.storefrontcloud.io) - this is a new feature or improvement for Vue Storefront. I've created branch from `develop` branch and want to merge it back to `develop`
- [x] RC version (https://next.storefrontcloud.io) - this is a stabilisation fix for Release Candidate of Vue Storefront. I've created branch from `release` branch and want to merge it back to `release`
- [ ] Stable version (https://demo.storefrontcloud.io) - this is an important fix for current stable version. I've created branch from `hotfix` or `master` branch and want to merge it back to `hotfix`

### Upgrade Notes and Changelog
- [x] No upgrade steps required (100% backward compatibility and no breaking changes)
- [x] I've updated the [Upgrade notes](https://github.com/DivanteLtd/vue-storefront/blob/develop/docs/guide/upgrade-notes/README.md) and [Changelog](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md) on how to port existing Vue Storefront sites with this new feature

**IMPORTANT NOTICE** - Remember to update `CHANGELOG.md` with description of your change

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/DivanteLtd/vue-storefront/blob/master/CONTRIBUTING.md)

### CODE REVIEW

File: core/modules/cart/helpers/productsEquals.ts [hunk 1]
```diff
@@ -20,14 +15,73 @@ const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean => {
   return areItemIdsDefined && product1ItemId === product2ItemId
 }
 
+// 'checksum' check
+const getChecksum = (product: CartItem) => {
```
Comment (Michal-Dziedzinski) [inline core/modules/cart/helpers/productsEquals.ts]:
```suggestion
const getChecksum = (product: CartItem) => product.checksum ? product.checksum : productChecksum(product)
```
mayby shorter form, wdyt?  :)

File: core/modules/cart/helpers/productsEquals.ts [hunk 2]
```diff
@@ -1,17 +1,12 @@
...
-  }
-  return productChecksum(product)
-}
+type ProductEqualCheckFn = (product1: CartItem, product2: CartItem) => boolean
 
+// 'id' check
 const getServerItemId = (product: CartItem): string | number =>
   product.server_item_id || product.item_id
-
-const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean => {
```
Comment (Michal-Dziedzinski) [inline core/modules/cart/helpers/productsEquals.ts]: Why types are removed?

File: core/modules/cart/helpers/productsEquals.ts [hunk 3]
```diff
@@ -20,14 +15,73 @@ const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean => {
...
+    return product.checksum
+  }
+  return productChecksum(product)
+}
 const isChecksumEquals = (product1: CartItem, product2: CartItem): boolean =>
   getChecksum(product1) === getChecksum(product2)
 
+// 'sku' check
+const isSkuEqual = (product1: CartItem, product2: CartItem): boolean =>
+  String(product1.sku) === String(product2.sku)
```
Comment (Michal-Dziedzinski) [inline core/modules/cart/helpers/productsEquals.ts]: Can be `sku` another type than string?

Comment (Michal-Dziedzinski) [inline core/modules/cart/helpers/productsEquals.ts]: Can be `sku` another type than string?

File: core/modules/cart/helpers/productsEquals.ts [hunk 4]
```diff
@@ -20,14 +15,73 @@ const isServerIdsEquals = (product1: CartItem, product2: CartItem): boolean => {
...
+ */
+const getCheckFn = (checkName: string): ProductEqualCheckFn => {
+  switch (checkName) {
+    case 'id': {
+      return isServerIdsEquals
+    }
+    case 'checksum': {
+      return isChecksumEquals
+    }
+    case 'sku': {
```
Comment (Michal-Dziedzinski) [inline core/modules/cart/helpers/productsEquals.ts]: shouldn't it be remove if it is the same as default?

**Comment by alinadivante** [comment]
@gibkigonzo 
Regarding #4104 
1. When I change color than photo is not updated
2. Do not close Edit mode pop-up window until the changed values [color or sizes] are updated.
Now they update with a delay ...
edit_mode

**Pull request merged**